options.go

```diff
@@ -15,9 +15,15 @@ func (fn optionFunc[K, V]) apply(opts *options[K, V]) {
 	fn(opts)
 }
 
+// CostCalcFunc is used to calculate the cost of the key and the item to be
+// inserted into the cache.
+type CostCalcFunc[K comparable, V any] func(key K, item V) uint64
```

Comment:
```suggestion
type CostFunc[K comparable, V any] func(item *Item[K, V]) uint64
```

Is there a reason why we can't have the actual `*Item[K, V]` as a parameter here? With it, the API would be more consistent and predictable.

Reply:
The only complication that I can see is in the `set()` method: when you are creating the `oldValue` variable, you might need to clone the whole item, otherwise the same updated item/pointer will be used when calculating old and new costs.

If you need to clone an item there, it might be a good idea to add a `Clone()` method to the item type (it should just recreate the item with the same fields while the item's mutex is locked -- it's the same logic as in any other exported Item method).

Reply:
hm... I don't think cloning an item is a good idea - it might be expensive. The idea I'm thinking of instead, is adding a new `cost` property to the `Item` struct and reuse it.

Reply:
The last commit includes a change to the signature of the CostFunc function as you suggested. Instead of holding a cost of each item, I moved the calculation of the old cost to a line before the item is updated for now.

Reply:
Indeed, I’d prefer avoiding the recalculation of old costs during updates and instead storing the calculated cost directly in the item. When a new item is created, it would invoke the `CostFunc` to compute and store the cost, and the same would happen during an update. However, this approach would require modifying the `NewItem` function, which would be a breaking change. Honestly, it also raises the question of why the `NewItem` function is public in the first place.

Reply:
> hm... I don't think cloning an item is a good idea - it might be expensive.

Yup, that's a good point.

> When a new item is created, it would invoke the CostFunc to compute and store the cost, and the same would happen during an update.

That would probably be more accurate as well, since the value inside the item might change after creation (i.e., the user would still have access to the value and they might change it from the outside).

> However, this approach would require modifying the NewItem function, which would be a breaking change. Honestly, it also raises the question of why the NewItem function is public in the first place?

Yes, that's a good question... For the longest time, this function was private and the reasons for that are similar to what you said/are trying to add. However, in #151 it was made public, which is not a bad change as it serves a purpose and has a use case, but it could have been implemented a bit better: just like the Set/Get methods, it should probably accept functional options as a parameter to update its internal fields. Unfortunately, we cannot change the signature of this function before the next major release, so we have two options:
- Create another `NewItem` function (e.g., `NewItemWithOpts`) and accept functional options, and then deprecate the old `NewItem`. It would also be removed in the next major release.
- Create a private `NewItem` alternative until `NewItemWithOpts` is added. I don't like this approach since it would undermine the whole purpose of #151. However, for now, to finish this feature, it might suffice. `NewItemWithOpts` could be added in a separate PR.

Reply:
@dadrus let me know when you'd like me to review your changes.

Reply:
Will do so. I'm pretty swamped this week, but hope to finalize the PR during the WE.

Most of the changes, we were discussing are already implemented. Some of the tests require however updates